entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
-These tests run tens of thousands of queries in a couple of minutes on a fast machine.
-This illustrates that using the real database system can be a performant solution.
-It is a myth that using some lighter-weight database is the only way to run tests quickly.
+Compared to running integration tests against your production database system, it's relatively easy to get started with SQLite, and so many users do. Unfortunately, the above limitations tend to eventually become problematic when unit testing EF Core applications, even if they don't seem to be at the beginning. As a result, we recommend either concentrating on integration tests, or implementing unit tests via the repository pattern as discussed below.
 
-That being said, what if for whatever reason you can't run tests against something close to your production database system?
-The next best choice is to use something with similar functionality.
-This usually means another relational database, for which [SQLite](https://sqlite.org/index.html) is the obvious choice.
+For information on how to use SQLite for unit testing, see the [unit testing page](xref:core/testing/unit-testing#sqlite-in-memory).
 
-SQLite is a good choice because:
+### InMemory as a database fake
 
-* It runs in-process with your application and so has low overhead.
-* It uses simple, automatically created files for databases, and so doesn't require database management.
-* It has an in-memory mode that avoids even the file creation.
+As an alternative to SQLite, EF Core also comes with an InMemory provider. Although this provider was originally designed to support internal testing of EF Core itself, some developers use it as a database fake when unit testing EF Core using EF Core. Doing so is **highly discouraged**: as a database fake, InMemory has the same issues as SQLite (see above), but in addition has the following additional limitations:
 
-However, remember that:
+* The InMemory provider generally supports less query types than the SQLite provider, since it isn't a relational database. More queries will fail or behave differently in comparison to your production database.
+* Transactions are not supported.
+* Raw SQL is completely unsupported. Compare this with SQLite, where it's possible to use raw SQL, as long as that SQL works in the same way on SQLite and your production database.
+* The InMemory provider has not been optimized for performance, and will generally work slower than SQLite in in-memory mode.
 
-* SQLite inevitably doesn't support everything that your production database system does.
-* SQLite will behave differently than your production database system for some queries.
+In summary, InMemory has all the disadvantages of SQLite, along with a few more - and offers no advantages in return. If you are looking for a simple, in-memory database fake, use SQLite instead of the InMemory provider; but consider using the repository pattern instead as described below.
```

Comment:
I think it has the advantage that its very easy to use.

It's interesting, but after reading these sections my conclusions are:
- Don't use another database as a fake unless what the database does is secondary to what you are testing. (In other words, it can be fine to easily get a list of customers and products that are then used to test some business logic.)
- If you're going to use a fake, then understand that is a fake and use the easiest thing possible, which might be in-memory.

Fundamentally, neither SQLite not the in-memory provider will behave the same as your production database, so _it doesn't make sense to use either of these_ for any testing where that is important. Further, it _is easy_ to use some form of your production database system, and _it doesn't have these issues_, so do that. It will always be better.

Reply:
We briefly discussed this offilne. Here are some points:

* I agree with yuor first bullet point: if you're using only very basic/universal database capabilities - simple read/write, no EF.Functions - using a fake (either InMemory or SQLite) may be sufficient; it's certainly easier to set up testing (e.g. in a CI pipeline) against a fake, and unlike repository doesn't impact your program architecture in an extreme way.
* However, on the second point, I do believe SQLite is advantageous because of the 4 points listed above: you get transactions, raw SQL (where the SQL works the same way as on your production database!), more queries translate and in a more similar way (because it's relational), and possibly speed. So within the (necessarily limited!) world of fake databases, SQLite does widen the scope of stuff you can unit-test without adopting a repository.
* Sure, there's that lifetime issue of SQLite in-memory, but that's a one-time startup thing you need to grok, and then it's OK. I wouldn't recommend against SQLite solely because of that.

> Fundamentally, neither SQLite not the in-memory provider will behave the same as your production database, so it doesn't make sense to use either of these for any testing where that is important. Further, it is easy to use some form of your production database system, and it doesn't have these issues, so do that. It will always be better.

So again, I think it's a matter of degree. For full testability/fidelity, I agree you need your real database, or a repository layer. But if you're willing to compromise on that to avoid real database setup, isolation issues, etc., and your application doesn't need EF.Functions.Whatever, then a fake can be reasonable, and some fakes (SQLite) are closer to your real database than others (InMemory).